### ISSUE TITLE
Obtain metrics for validation dataset during training 

### ISSUE THREAD

**Issue by mmeendez8** [open-mmlab/mmdetection#2409]
**Describe the feature**
When validation flag is selected in `api/train.py`, `EvaluationHook` gets inmediately registered (see [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/apis/train.py#L236)). Nevertheless, **metrics are only computed over training dataset** once each training epoch is done, since EvalHook only implements `after_train_epoch` method. 

I consider it would be more interesting **to obtain those metrics over validation dataset** and it's not clear to why this has no been activated.

**Motivation**
It's good to use validation data to understand how your model is learning with the number of epochs. Losses are crucial for this and you can infer a lot by comparing training and validation curves. But some times loss values are not enough for the problem you are dealing with, slight changes in losses may end up on large differences in model performance. 

**Additional context**

1. One plausible solution for this problem would be to change `EvalHook` replacing its `after_train_epoch` with `after_epoch`method:

```
class EvalHook(Hook):
...
def after_epoch(self, runner):
        if not self.every_n_epochs(runner, self.interval):
            return
        from mmdet.apis import single_gpu_test
        results = single_gpu_test(runner.model, self.dataloader, show=False)
        self.evaluate(runner, results)

```

**Drawbacks**: This is far from optimal since we would be feeding validation data twice, one per computing loss values and other for validation metrics.  

2. Adapt code to save batch predictions after each iteration of `runner.val()` method. This requires a large number of modifications in the source code:
    - Modify `apis.train.batch_processor()`  and detectors `forward_train()' method to return bboxes when required.
    - Evaluate and accumulate bbox results per batch
    - Compute metrics from accumulated batch results

**Drawbacks**: It would require a large refactor of code base. Runner module is affected so mmcv package would be also afected.

We are interested in this feature and willing to develop it. I'm open this discussion to serve as a debate about the changes to be made

**Comment by hellock** [comment]
Evaluation hooks use the validation dataset instead of training dataset, see [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/apis/train.py#L175) for details. `after_train_epoch()` only indicates that it is triggered after each (or several) training epochs, and `runner.val()` is never called because we specify `workflow=[('train', 1)]` in config files.

**Comment by mmeendez8** [comment]
I see, thanks your that explanation @hellock, we missed that point!

**Comment by Zero-2** [comment]
is there any way that i can compute the metrics on trianing set?

**Comment by Zero-2** [comment]
is there any way that i can compute the metrics on training set?
